File: ts/src/lib/emitters/sound.ts
import { WritableByteStream } from "@open-flash/stream";
import { Incident } from "incident";
import { Uint2, Uint4, Uint8 } from "semantic-types";
import { AudioCodingFormat } from "swf-types/sound/audio-coding-format";
import { SoundEnvelope } from "swf-types/sound/sound-envelope";
import { SoundInfo } from "swf-types/sound/sound-info";
import { SoundRate } from "swf-types/sound/sound-rate";

export function audioCodingFormatToCode(value: AudioCodingFormat): Uint4 {
  switch (value) {
    case AudioCodingFormat.UncompressedNativeEndian:
      return 0;
    case AudioCodingFormat.Adpcm:
      return 1;
    case AudioCodingFormat.Mp3:
      return 2;
    case AudioCodingFormat.UncompressedLittleEndian:
      return 3;
    case AudioCodingFormat.Nellymoser16:
      return 4;
    case AudioCodingFormat.Nellymoser8:
      return 5;
    case AudioCodingFormat.Nellymoser:
      return 6;
    case AudioCodingFormat.Speex:
      return 11;
    default:
      throw new Incident("UnexpectedAudioCodingFormat");
  }
}

export function soundRateToCode(value: SoundRate): Uint2 {
  switch (value) {
    case 5500:
      return 0;
    case 11000:
      return 1;
    case 22000:
      return 2;
    case 44000:
      return 3;
    default:
      throw new Incident("UnexpectedSoundRate");
  }
}

export function emitSoundInfo(byteStream: WritableByteStream, value: SoundInfo): void {
  const hasInPoint: boolean = value.inPoint !== undefined;
  const hasOutPoint: boolean = value.outPoint !== undefined;
  const hasLoops: boolean = value.loopCount !== undefined;
  const hasEnvelope: boolean = value.envelopeRecords !== undefined;
  const syncNoMultiple: boolean = value.syncNoMultiple;
  const syncStop: boolean = value.syncStop;

  const flags: Uint8 = 0
    | (hasInPoint ? 1 << 0 : 0)
    | (hasOutPoint ? 1 << 1 : 0)
    | (hasLoops ? 1 << 2 : 0)
    | (hasEnvelope ? 1 << 3 : 0)
    | (syncNoMultiple ? 1 << 4 : 0)
    | (syncStop ? 1 << 5 : 0);
  // Skip bits [6, 7]
  byteStream.writeUint8(flags);

  if (value.inPoint !== undefined) {
    byteStream.writeUint32LE(value.inPoint);
  }
  if (value.outPoint !== undefined) {
    byteStream.writeUint32LE(value.outPoint);
  }
  if (value.loopCount !== undefined) {
    byteStream.writeUint16LE(value.loopCount);
  }
  if (value.envelopeRecords !== undefined) {
    emitSoundEnvelope(byteStream, value.envelopeRecords);
  }
}

export function emitSoundEnvelope(
  byteStream: WritableByteStream,
  value: readonly SoundEnvelope[],
): void {
  byteStream.writeUint8(value.length);
  for (const record of value) {
    byteStream.writeUint32LE(record.pos44);
    byteStream.writeUint16LE(record.leftLevel);
    byteStream.writeUint16LE(record.rightLevel);
  }
}

File: rs/src/sound.rs
use std::convert::TryInto;
use std::io;

use swf_types as ast;

use crate::primitives::{emit_le_u16, emit_le_u32, emit_u8};

pub(crate) fn audio_coding_format_to_code(value: ast::AudioCodingFormat) -> u8 {
  match value {
    ast::AudioCodingFormat::UncompressedNativeEndian => 0,
    ast::AudioCodingFormat::Adpcm => 1,
    ast::AudioCodingFormat::Mp3 => 2,
    ast::AudioCodingFormat::UncompressedLittleEndian => 3,
    ast::AudioCodingFormat::Nellymoser16 => 4,
    ast::AudioCodingFormat::Nellymoser8 => 5,
    ast::AudioCodingFormat::Nellymoser => 6,
    ast::AudioCodingFormat::Speex => 11,
  }
}

pub(crate) fn sound_rate_to_code(value: ast::SoundRate) -> u8 {
  match value {
    ast::SoundRate::SoundRate5500 => 0,
    ast::SoundRate::SoundRate11000 => 1,
    ast::SoundRate::SoundRate22000 => 2,
    ast::SoundRate::SoundRate44000 => 3,
  }
}

pub(crate) fn emit_sound_info<W: io::Write>(writer: &mut W, value: &ast::SoundInfo) -> io::Result<()> {
  let has_in_point = value.in_point.is_some();
  let has_out_point = value.out_point.is_some();
  let has_loops = value.loop_count.is_some();
  let has_envelope = value.envelope_records.is_some();
  let sync_no_multiple = value.sync_no_multiple;
  let sync_stop = value.sync_stop;

  #[allow(clippy::identity_op)]
  let flags: u8 = 0
    | (if has_in_point { 1 << 0 } else { 0 })
    | (if has_out_point { 1 << 1 } else { 0 })
    | (if has_loops { 1 << 2 } else { 0 })
    | (if has_envelope { 1 << 3 } else { 0 })
    | (if sync_no_multiple { 1 << 4 } else { 0 })
    | (if sync_stop { 1 << 5 } else { 0 });
  // Skip bits [6, 7]
  emit_u8(writer, flags)?;

  if let Some(in_point) = value.in_point {
    emit_le_u32(writer, in_point)?;
  }
  if let Some(out_point) = value.out_point {
    emit_le_u32(writer, out_point)?;
  }
  if let Some(loop_count) = value.loop_count {
    emit_le_u16(writer, loop_count)?;
  }
  if let Some(ref envelope) = &value.envelope_records {
    emit_sound_envelope(writer, envelope)?;
  }

  Ok(())
}

pub(crate) fn emit_sound_envelope<W: io::Write>(writer: &mut W, value: &[ast::SoundEnvelope]) -> io::Result<()> {
  emit_u8(writer, value.len().try_into().unwrap())?;
  for record in value {
    emit_le_u32(writer, record.pos44)?;
    emit_le_u16(writer, record.left_level)?;
    emit_le_u16(writer, record.right_level)?;
  }

  Ok(())
}
